<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Interoptability between MAST and SingleCellExperiment-derived packages. • MAST</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Interoptability between MAST and SingleCellExperiment-derived packages.">
<meta property="og:description" content="MAST">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.21.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/MAITAnalysis.html">Using MAST with RNASeq: MAIT Analysis.</a>
    </li>
    <li>
      <a href="../articles/MAST-Intro.html">MAST Intro</a>
    </li>
    <li>
      <a href="../articles/MAST-interoperability.html">Interoptability between MAST and SingleCellExperiment-derived packages.</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/RGLab/MAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="MAST-interoperability_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Interoptability between MAST and SingleCellExperiment-derived packages.</h1>
                        <h4 data-toc-skip class="author">Andrew McDavid</h4>
            
            <h4 data-toc-skip class="date">2022-02-01</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/RGLab/MAST/blob/HEAD/vignettes/MAST-interoperability.Rmd" class="external-link"><code>vignettes/MAST-interoperability.Rmd</code></a></small>
      <div class="hidden name"><code>MAST-interoperability.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>As a <code>SingleCellExperiment</code>-derived package, <code>MAST</code> can easily be inserted into workflows with packages such as <code>scran</code>, <code>scater</code>, <code>zinbwave</code>, <code>SCnorm</code> and others. Moreover, subclassing SingleCellExperiment/SummarizedExperiment provides a flexible abstraction for the <code>assay</code> that contains the actual expression data. It can use sparse <code>Matrix</code> and HDF5 as backends to save memory.</p>
<p>To use MAST with such packages, you just need to upcast the <code>SingleCellExperiment</code> to MAST’s subclass <code>SingleCellAssay</code> with the function <code>SceToSingleCellAssay</code> that handles the coercion and checks the object for validity. Going the other direction, generally <code>SingleCellAssay</code>s should work in packages that use <code>SingleCellExperiment</code>, but if in doubt you could down-cast with <code>as(sca, 'SingleCellExperiment')</code>.</p>
<div class="section level3">
<h3 id="log-transformation-is-expected-in-mast">Log-transformation is expected in MAST<a class="anchor" aria-label="anchor" href="#log-transformation-is-expected-in-mast"></a>
</h3>
<p>The main gotcha in all this is that some SingleCellExperiment-derived packages assume integer counts have been provided, while MAST assumes that log-transformed approximately scale-normalized data is provided. We find that MAST performs best with log-transformed, scale-normalized data that has been thresholded, such as <span class="math inline">\(\log_2(\text{transcripts per million} + 1)\)</span>.</p>
<p>We address this by:</p>
<ul>
<li>testing for log-like data for objects up-cast to <code>SingleCellAssay</code>
</li>
<li>explicitly naming the slot of the <code>assay</code> containing such putatively log-like data</li>
<li>by default operating on the slot with such log-like data</li>
</ul>
</div>
<div class="section level3">
<h3 id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h3>
<p>In what follows, we show an example of using <code>scater</code> to plot some QC metrics, <code>SCnorm</code> to normalize data, and, and conversion to a <code>Seurat</code> object.</p>
</div>
</div>
<div class="section level2">
<h2 id="from-mast-to-scater">From MAST to Scater<a class="anchor" aria-label="anchor" href="#from-mast-to-scater"></a>
</h2>
<p>Scater <span class="citation">McCarthy et al. (2017)</span> is a package that provides functions for QC, normalization and visualization of single cell RNAseq data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/RGLab/MAST/" class="external-link">MAST</a></span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Loading required package: SingleCellExperiment</span></code></pre>
<pre><code><span class="co">## Loading required package: SummarizedExperiment</span></code></pre>
<pre><code><span class="co">## Loading required package: MatrixGenerics</span></code></pre>
<pre><code><span class="co">## Loading required package: matrixStats</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'MatrixGenerics'</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:matrixStats':</span>
<span class="co">## </span>
<span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span>
<span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span>
<span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span>
<span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span>
<span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span>
<span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span>
<span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span>
<span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span>
<span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span>
<span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span>
<span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span>
<span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span>
<span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span>
<span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span>
<span class="co">##     rowWeightedSds, rowWeightedVars</span></code></pre>
<pre><code><span class="co">## Loading required package: GenomicRanges</span></code></pre>
<pre><code><span class="co">## Loading required package: stats4</span></code></pre>
<pre><code><span class="co">## Loading required package: BiocGenerics</span></code></pre>
<pre><code><span class="co">## Loading required package: parallel</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'BiocGenerics'</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:parallel':</span>
<span class="co">## </span>
<span class="co">##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,</span>
<span class="co">##     clusterExport, clusterMap, parApply, parCapply, parLapply,</span>
<span class="co">##     parLapplyLB, parRapply, parSapply, parSapplyLB</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:stats':</span>
<span class="co">## </span>
<span class="co">##     IQR, mad, sd, var, xtabs</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:base':</span>
<span class="co">## </span>
<span class="co">##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,</span>
<span class="co">##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,</span>
<span class="co">##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,</span>
<span class="co">##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,</span>
<span class="co">##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,</span>
<span class="co">##     union, unique, unsplit, which.max, which.min</span></code></pre>
<pre><code><span class="co">## Loading required package: S4Vectors</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'S4Vectors'</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:base':</span>
<span class="co">## </span>
<span class="co">##     expand.grid, I, unname</span></code></pre>
<pre><code><span class="co">## Loading required package: IRanges</span></code></pre>
<pre><code><span class="co">## Loading required package: GenomeInfoDb</span></code></pre>
<pre><code><span class="co">## Loading required package: Biobase</span></code></pre>
<pre><code><span class="co">## Welcome to Bioconductor</span>
<span class="co">## </span>
<span class="co">##     Vignettes contain introductory material; view with</span>
<span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span>
<span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></code></pre>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'Biobase'</span></code></pre>
<pre><code><span class="co">## The following object is masked from 'package:MatrixGenerics':</span>
<span class="co">## </span>
<span class="co">##     rowMedians</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:matrixStats':</span>
<span class="co">## </span>
<span class="co">##     anyMissing, rowMedians</span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_chunk.html" class="external-link">opts_chunk</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>message <span class="op">=</span> <span class="cn">FALSE</span>,error <span class="op">=</span> <span class="cn">FALSE</span>,warning <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">maits</span>, package<span class="op">=</span><span class="st">'MAST'</span><span class="op">)</span>
<span class="va">unlog</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">ceiling</a></span><span class="op">(</span><span class="fl">2</span><span class="op">^</span><span class="va">x</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span>
<span class="va">sca_raw</span> <span class="op">=</span>  <span class="fu"><a href="../reference/FromMatrix.html">FromMatrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">maits</span><span class="op">$</span><span class="va">expressionmat</span><span class="op">)</span>, <span class="va">maits</span><span class="op">$</span><span class="va">cdat</span>, <span class="va">maits</span><span class="op">$</span><span class="va">fdat</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Assuming data assay in position 1, with name et is log-transformed.</span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assays</span><span class="op">(</span><span class="va">sca_raw</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span> <span class="op">=</span> <span class="fu">unlog</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">sca_raw</span><span class="op">)</span><span class="op">)</span>
<span class="fu">assayNames</span><span class="op">(</span><span class="va">sca_raw</span><span class="op">)</span></code></pre></div>
<p>Here we make an object with assays <code>counts</code> and <code>et</code>. By default, <code>MAST</code> will operate on the <code>et</code> assay, but scran wants count-like data for some of its QC. The <code>et</code> data are log2 + 1 transcripts per million (TPM), as output by RSEM.</p>
<p>We could specify the assay name at creation with <code>sca_raw = FromMatrix(list(logTPM = t(maits$expressionmat)), maits$cdat, maits$fdat)</code> or rename an object that contains appropriately transformed data with <code>assayNames(sca_raw) = c('logTPM', 'counts')</code>.</p>
<p>Before calling <code>scater</code> functionality, you might pause to consider if some features should belong in special <code>control</code> sets, such as mitochrondial genes, or spike-ins.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span>
<span class="va">sca_raw</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQC</a></span><span class="op">(</span><span class="va">sca_raw</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">sca_raw</span>, y<span class="op">=</span><span class="st">"detected"</span>, x<span class="op">=</span><span class="st">"total"</span><span class="op">)</span></code></pre></div>
<p><img src="MAST-interoperability_files/figure-html/scaterQC-1.png" width="700"></p>
<p>Evidently some features were filtered, so not all cells contain 1 million counts.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sca_raw</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sca_raw</span>, ncomponents<span class="op">=</span><span class="fl">5</span>, exprs_values <span class="op">=</span> <span class="st">'et'</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sca_raw</span>, dimred <span class="op">=</span> <span class="st">'PCA'</span>, colour_by <span class="op">=</span> <span class="st">'condition'</span><span class="op">)</span></code></pre></div>
<p><img src="MAST-interoperability_files/figure-html/unnamed-chunk-1-1.png" width="700"> We can also run a PCA.</p>
<div class="section level3">
<h3 id="from-scater-to-mast">From scater to MAST<a class="anchor" aria-label="anchor" href="#from-scater-to-mast"></a>
</h3>
<p>Since scater uses <code>SingleCellExperiment</code> objects, the only here consideration is making sure <code>MAST</code> can find log-like data, and possibly thresholding the data.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R">  <span class="va">example_sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/mockSCE.html" class="external-link">mockSCE</a></span><span class="op">(</span><span class="op">)</span> 

<span class="va">example_sce</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span>
<span class="va">sca</span> <span class="op">=</span> <span class="fu"><a href="../reference/SceToSingleCellAssay.html">SceToSingleCellAssay</a></span><span class="op">(</span><span class="va">example_sce</span><span class="op">)</span></code></pre></div>
<p>Here we coerce <code>example_sce</code> to be a SingleCellAssay object.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/zlm.html">zlm</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">Treatment</span>, sca <span class="op">=</span> <span class="va">sca</span>, exprs_value <span class="op">=</span> <span class="st">'logcounts'</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Fitted zlm on 2000 genes and 200 cells.</span>
<span class="co">##  Using BayesGLMlike ~ Treatment</span></code></pre>
<p>We test for differential expression with regards to <code>Treatment</code> and explicitly indicate the <code>logcounts</code> slot will be used. Methods in MAST will operate on the default slice returned by <code>assay</code>, which has been over-ridden to return log-like data: the default slice is the first assay whose name, as given by <code>assayNames(x)</code>, matches the first element in the sequence <code>c('thresh', 'et', 'Et', 'lCount', 'logTPM', 'logCounts', 'logcounts')</code>. So in the case of <code>sca</code>, even if <code>exprs_value</code> was not specified, the <code>logcounts</code> slot would have been used, even though it comes second in <code>assayNames(sca)</code>:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">assayNames</span><span class="op">(</span><span class="va">sca</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "counts"    "logcounts"</span></code></pre>
</div>
<div class="section level3">
<h3 id="sparse-matrix-and-hdf5-support">Sparse matrix and HDF5 support<a class="anchor" aria-label="anchor" href="#sparse-matrix-and-hdf5-support"></a>
</h3>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Matrix.R-forge.R-project.org/" class="external-link">Matrix</a></span><span class="op">)</span>
<span class="va">sca_sparse</span> <span class="op">=</span> <span class="fu"><a href="../reference/FromMatrix.html">FromMatrix</a></span><span class="op">(</span>
    exprsArray <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>et <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Matrix/man/Matrix.html" class="external-link">Matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">maits</span><span class="op">$</span><span class="va">expressionmat</span><span class="op">)</span>, sparse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>,
    cData <span class="op">=</span> <span class="va">maits</span><span class="op">$</span><span class="va">cdat</span>, fData <span class="op">=</span> <span class="va">maits</span><span class="op">$</span><span class="va">fdat</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">sca_sparse</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "dgCMatrix"</span>
<span class="co">## attr(,"package")</span>
<span class="co">## [1] "Matrix"</span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">regular_time</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="fu"><a href="../reference/zlm.html">zlm</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">condition</span>, sca <span class="op">=</span> <span class="va">sca_raw</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span>
<span class="va">sparse_time</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="fu"><a href="../reference/zlm.html">zlm</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">condition</span>, sca <span class="op">=</span> <span class="va">sca_sparse</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>There is no complication to providing a sparse matrix.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/DelayedArray" class="external-link">DelayedArray</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/HDF5Array" class="external-link">HDF5Array</a></span><span class="op">)</span>
<span class="va">hd5_dat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedMatrix-utils.html" class="external-link">t</a></span><span class="op">(</span><span class="va">maits</span><span class="op">$</span><span class="va">expressionmat</span><span class="op">)</span>, <span class="st">'HDF5Array'</span><span class="op">)</span>
<span class="fu">DelayedArray</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/showtree.html" class="external-link">seed</a></span><span class="op">(</span><span class="va">hd5_dat</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## An object of class "HDF5ArraySeed"</span>
<span class="co">## Slot "filepath":</span>
<span class="co">## [1] "/tmp/RtmpNbzSji/HDF5Array_dump/auto00001.h5"</span>
<span class="co">## </span>
<span class="co">## Slot "name":</span>
<span class="co">## [1] "/HDF5ArrayAUTO00001"</span>
<span class="co">## </span>
<span class="co">## Slot "as_sparse":</span>
<span class="co">## [1] FALSE</span>
<span class="co">## </span>
<span class="co">## Slot "type":</span>
<span class="co">## [1] NA</span>
<span class="co">## </span>
<span class="co">## Slot "dim":</span>
<span class="co">## [1] 16302    96</span>
<span class="co">## </span>
<span class="co">## Slot "chunkdim":</span>
<span class="co">## [1] 13031    76</span>
<span class="co">## </span>
<span class="co">## Slot "first_val":</span>
<span class="co">## [1] 0</span></code></pre>
<p>Write <code>sc_example_counts</code> to disk as an <code>HDF5Array</code></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sca_delay</span> <span class="op">=</span> <span class="fu"><a href="../reference/FromMatrix.html">FromMatrix</a></span><span class="op">(</span>
    exprsArray <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>et <span class="op">=</span> <span class="va">hd5_dat</span><span class="op">)</span>,
     cData <span class="op">=</span> <span class="va">maits</span><span class="op">$</span><span class="va">cdat</span>, fData <span class="op">=</span> <span class="va">maits</span><span class="op">$</span><span class="va">fdat</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">sca_sparse</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "dgCMatrix"</span>
<span class="co">## attr(,"package")</span>
<span class="co">## [1] "Matrix"</span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">hd5_time</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="fu"><a href="../reference/zlm.html">zlm</a></span><span class="op">(</span> <span class="op">~</span> <span class="va">condition</span>, sca <span class="op">=</span> <span class="va">sca_delay</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span>,<span class="op">]</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Nor is there a complication to using HDF5-backed stores.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Dense'</span>, <span class="st">'Sparse'</span>, <span class="st">'HDF5'</span><span class="op">)</span>, <span class="st">'user time(s)'</span> <span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span> <span class="va">regular_time</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">sparse_time</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">hd5_time</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>, check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">method</th>
<th align="right">user time(s)</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Dense</td>
<td align="right">0.519</td>
</tr>
<tr class="even">
<td align="left">Sparse</td>
<td align="right">0.623</td>
</tr>
<tr class="odd">
<td align="left">HDF5</td>
<td align="right">12.724</td>
</tr>
</tbody>
</table>
<p>Dense storage is generally fastest, followed by the sparse storage. HDF5 is often slowest, but if your data doesn’t fit in memory, you don’t really have any other choice. The linear models underlying MAST don’t have any special provision for big <span class="math inline">\(n\)</span> data, and will tend to linearly (or worse) in the number of cells. So performance may be an issue even if they data do fit in memory.</p>
</div>
</div>
<div class="section level2">
<h2 id="mast-and-zinb-wave">MAST and ZINB-wave<a class="anchor" aria-label="anchor" href="#mast-and-zinb-wave"></a>
</h2>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">zinbwave</span><span class="op">)</span>
<span class="va">feature_var</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-utils.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sca_raw</span><span class="op">)</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span>
<span class="va">sca_top500</span> <span class="op">=</span> <span class="va">sca_raw</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/rank.html" class="external-link">rank</a></span><span class="op">(</span><span class="op">-</span><span class="va">feature_var</span><span class="op">)</span><span class="op">&lt;=</span><span class="fl">500</span>,<span class="op">]</span>
<span class="va">zw</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/zinbwave/man/zinbwave.html" class="external-link">zinbwave</a></span><span class="op">(</span>Y <span class="op">=</span> <span class="va">sca_top500</span>, X <span class="op">=</span> <span class="st">'~1'</span>, which_assay <span class="op">=</span> <span class="st">'counts'</span>, K <span class="op">=</span> <span class="fl">2</span>, normalizedValues <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Run zinbwave. To speed things, we take the top 500 most variable genes.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rd</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">zw</span>, <span class="st">'PCA'</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">zw</span>, <span class="st">'zinbwave'</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">zw</span><span class="op">)</span><span class="op">)</span>
<span class="fu">GGally</span><span class="fu">::</span><span class="fu"><a href="https://ggobi.github.io/ggally/reference/ggpairs.html" class="external-link">ggpairs</a></span><span class="op">(</span><span class="va">rd</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'PC1'</span>, <span class="st">'PC2'</span>, <span class="st">'W1'</span>, <span class="st">'W2'</span><span class="op">)</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">condition</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="MAST-interoperability_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="section level3">
<h3 id="using-mast-to-characterizing-genes-that-drive-the-factors">Using MAST to characterizing genes that drive the factors<a class="anchor" aria-label="anchor" href="#using-mast-to-characterizing-genes-that-drive-the-factors"></a>
</h3>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">zw</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/DelayedArray/man/DelayedArray-utils.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">zw</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">zw</span>, <span class="st">'zinbwave'</span><span class="op">)</span><span class="op">)</span>
<span class="va">zw</span> <span class="op">=</span> <span class="fu"><a href="../reference/SceToSingleCellAssay.html">SceToSingleCellAssay</a></span><span class="op">(</span><span class="va">zw</span><span class="op">)</span>
<span class="va">zz</span> <span class="op">=</span> <span class="fu"><a href="../reference/zlm.html">zlm</a></span><span class="op">(</span><span class="op">~</span><span class="va">W1</span> <span class="op">+</span> <span class="va">W2</span>, sca <span class="op">=</span> <span class="va">zw</span>, exprs_values <span class="op">=</span> <span class="st">'et'</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ss</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/NMF/man/assess.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">zz</span><span class="op">)</span>
<span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ss</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Fitted zlm with top 2 genes per contrast: ( log fold change Z-score ) primerid W1 W2<br>
2197 2.6 -4.1<em> 3002 11.3</em> -0.1 3312 5.2 -4.2<em> 3458 12.4</em> -0.1</p>
<table class="table">
<thead><tr class="header">
<th align="left">primerid</th>
<th align="left">W1</th>
<th align="left">W2</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">2197</td>
<td align="left">2.6</td>
<td align="left">-4.1*</td>
</tr>
<tr class="even">
<td align="left">3002</td>
<td align="left">11.3*</td>
<td align="left">-0.1</td>
</tr>
<tr class="odd">
<td align="left">3312</td>
<td align="left">5.2</td>
<td align="left">-4.2*</td>
</tr>
<tr class="even">
<td align="left">3458</td>
<td align="left">12.4*</td>
<td align="left">-0.1</td>
</tr>
</tbody>
</table>
<p>These are log-fold changes in the top few changes associated with factors 1 and 2.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span>
<span class="va">top5</span> <span class="op">=</span> <span class="va">ss</span><span class="op">$</span><span class="va">datatable</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="../reference/MAST-defunct.html">filter</a></span><span class="op">(</span><span class="va">component</span><span class="op">==</span><span class="st">'logFC'</span>, <span class="va">contrast</span> <span class="op"><a href="https://Rdatatable.gitlab.io/data.table/reference/like.html" class="external-link">%like%</a></span> <span class="st">'W'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">z</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">head</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">zw</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/as.data.table.html" class="external-link">as.data.table</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
<span class="va">dat</span> <span class="op">=</span> <span class="va">zw</span><span class="op">[</span><span class="va">top5</span><span class="op">$</span><span class="va">primerid</span>,<span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="st">'data.table'</span><span class="op">)</span>
<span class="va">dat</span> <span class="op">=</span> <span class="va">dat</span><span class="op">[</span>,<span class="op">!</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">dat</span><span class="op">)</span><span class="op">)</span>,with <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span>
<span class="va">plt</span> <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">dat</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">W1</span>, color <span class="op">=</span> <span class="va">condition</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">symbolid</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plt</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">et</span><span class="op">)</span></code></pre></div>
<p><img src="MAST-interoperability_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>Expression on “Et” scale (<span class="math inline">\(\log_2( TPM + 1)\)</span>)</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">plt</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">normalizedValues</span><span class="op">)</span></code></pre></div>
<p><img src="MAST-interoperability_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>Normalized expression from factor model</p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references hanging-indent">
<div id="ref-scater">
<p>McCarthy, Davis J., Kieran R. Campbell, Aaron T. L. Lun, and Quin F. Willis. 2017. “Scater: Pre-Processing, Quality Control, Normalisation and Visualisation of Single-Cell RNA-Seq Data in R.” <em>Bioinformatics</em> 33 (8): 1179–86. <a href="https://doi.org/10.1093/bioinformatics/btw777" class="external-link">https://doi.org/10.1093/bioinformatics/btw777</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Andrew McDavid, Greg Finak, Masanao Yajima.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
