<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ MAST</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MAST</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.21.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/MAITAnalysis.html">Using MAST with RNASeq: MAIT Analysis.</a>
    </li>
    <li>
      <a href="../articles/MAST-Intro.html">MAST Intro</a>
    </li>
    <li>
      <a href="../articles/MAST-interoperability.html">Interoptability between MAST and SingleCellExperiment-derived packages.</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/RGLab/MAST/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="loading-transforming-and-summarizing">Loading, transforming and summarizing <a href="#loading-transforming-and-summarizing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Functions to load data, add covariates or annotations, or provide high-level summaries.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="FromFlatDF.html">FromFlatDF()</a></code> </p>
        </td>
        <td><p>Construct a SingleCellAssay (or derived subclass) from a `flat` (melted) data.frame/data.table</p></td>
      </tr><tr><td>
          <p><code><a href="FromMatrix.html">FromMatrix()</a></code> </p>
        </td>
        <td><p>Construct a SingleCellAssay from a matrix or array of expression</p></td>
      </tr><tr><td>
          <p><code><a href="SceToSingleCellAssay.html">SceToSingleCellAssay()</a></code> </p>
        </td>
        <td><p>Coerce a SingleCellExperiment to some class defined in MAST</p></td>
      </tr><tr><td>
          <p><code><a href="convertMASTClassicToSingleCellAssay.html">convertMASTClassicToSingleCellAssay()</a></code> </p>
        </td>
        <td><p>Convert a MASTClassic SingleCellAssay</p></td>
      </tr><tr><td>
          <p><code><a href="defaultAssay.html">magic_assay_names()</a></code> <code><a href="defaultAssay.html">assay_idx()</a></code> <code><a href="defaultAssay.html">assay(<i>&lt;SingleCellAssay&gt;</i>,<i>&lt;missing&gt;</i>)</a></code> </p>
        </td>
        <td><p>Default <code>assay</code> returned</p></td>
      </tr><tr><td>
          <p><code><a href="melt.SingleCellAssay.html">melt(<i>&lt;SingleCellAssay&gt;</i>)</a></code> </p>
        </td>
        <td><p>"Melt" a <code>SingleCellAssay</code> matrix</p></td>
      </tr><tr><td>
          <p><code><a href="zlm.html">zlm()</a></code> </p>
        </td>
        <td><p>Zero-inflated regression for SingleCellAssay</p></td>
      </tr><tr><td>
          <p><code><a href="freq.html">freq()</a></code> <code><a href="freq.html">condmean()</a></code> <code><a href="freq.html">condSd()</a></code> <code><a href="freq.html">numexp()</a></code> </p>
        </td>
        <td><p>Summary statistics for genes in an experiment</p></td>
      </tr><tr><td>
          <p><code><a href="getConcordance.html">getConcordance()</a></code> <code><a href="getConcordance.html">getwss()</a></code> <code><a href="getConcordance.html">getss()</a></code> <code><a href="getConcordance.html">getrc()</a></code> </p>
        </td>
        <td><p>Get the concordance between two experiments</p></td>
      </tr><tr><td>
          <p><code><a href="read.fluidigm.html">read.fluidigm()</a></code> </p>
        </td>
        <td><p>Reads a Fluidigm Biomark (c. 2011) raw data file (or set of files)</p></td>
      </tr><tr><td>
          <p><code><a href="primerAverage.html">primerAverage()</a></code> </p>
        </td>
        <td><p>Average expression values for duplicated/redundant genes</p></td>
      </tr><tr><td>
          <p><code><a href="MAST-defunct.html">MAST-defunct</a></code> </p>
        </td>
        <td><p>Defunct functions in package `MAST`</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="thresholding">Thresholding <a href="#thresholding" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Adaptive thresholding of background noise.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="thresholdSCRNACountMatrix.html">thresholdSCRNACountMatrix()</a></code> </p>
        </td>
        <td><p>Threshold a count matrix using an adaptive threshold.</p></td>
      </tr><tr><td>
          <p><code><a href="summary.thresholdSCRNACountMatrix.html">summary(<i>&lt;thresholdSCRNACountMatrix&gt;</i>)</a></code> <code><a href="summary.thresholdSCRNACountMatrix.html">print(<i>&lt;summaryThresholdSCRNA&gt;</i>)</a></code> </p>
        </td>
        <td><p>Summarize the effect of thresholding</p></td>
      </tr><tr><td>
          <p><code><a href="plot.thresholdSCRNACountMatrix.html">plot(<i>&lt;thresholdSCRNACountMatrix&gt;</i>)</a></code> </p>
        </td>
        <td><p>Plot cutpoints and densities for thresholding</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="differential-expression-testing-with-hurdle-model">Differential Expression Testing with Hurdle Model <a href="#differential-expression-testing-with-hurdle-model" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Fit a Hurdle linear model to test for zero-inflated differential expression.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="zlm.html">zlm()</a></code> </p>
        </td>
        <td><p>Zero-inflated regression for SingleCellAssay</p></td>
      </tr><tr><td>
          <p><code><a href="LMERlike-class.html">update(<i>&lt;LMERlike&gt;</i>)</a></code> <code><a href="LMERlike-class.html">vcov(<i>&lt;LMERlike&gt;</i>)</a></code> <code><a href="LMERlike-class.html">coef(<i>&lt;LMERlike&gt;</i>)</a></code> <code><a href="LMERlike-class.html">logLik(<i>&lt;LMERlike&gt;</i>)</a></code> </p>
        </td>
        <td><p>Wrapper for lmer/glmer</p></td>
      </tr><tr><td>
          <p><code><a href="BayesGLMlike-class.html">BayesGLMlike-class</a></code> </p>
        </td>
        <td><p>Wrapper for bayesian GLM</p></td>
      </tr><tr><td>
          <p><code><a href="print.summaryZlmFit.html">print(<i>&lt;summaryZlmFit&gt;</i>)</a></code> </p>
        </td>
        <td><p>Print summary of a ZlmFit</p></td>
      </tr><tr><td>
          <p><code><a href="ebayes.html">ebayes()</a></code> </p>
        </td>
        <td><p>Estimate hyperparameters for hierarchical variance model for continuous component</p></td>
      </tr><tr><td>
          <p><code><a href="defaultPrior.html">defaultPrior()</a></code> </p>
        </td>
        <td><p>Initialize a prior to be used a prior for BayeGLMlike/BayesGLMlike2</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="specifying-hypothesis-and-contrasts-to-test">Specifying hypothesis and contrasts to test <a href="#specifying-hypothesis-and-contrasts-to-test" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Testing linear functions of coefficients.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="Hypothesis.html">Hypothesis()</a></code> </p>
        </td>
        <td><p>Describe a linear model hypothesis to be tested</p></td>
      </tr><tr><td>
          <p><code><a href="lrTest.html">lrTest()</a></code> </p>
        </td>
        <td><p>Run a likelihood-ratio test</p></td>
      </tr><tr><td>
          <p><code><a href="waldTest.html">waldTest()</a></code> </p>
        </td>
        <td><p>Run a Wald test</p></td>
      </tr><tr><td>
          <p><code><a href="lrTest-ZlmFit-character-method.html">lrTest(<i>&lt;ZlmFit&gt;</i>,<i>&lt;character&gt;</i>)</a></code> </p>
        </td>
        <td><p>Likelihood ratio test</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="low-level-manipulation-of-fitted-hurdle-models">Low-level manipulation of fitted hurdle models <a href="#low-level-manipulation-of-fitted-hurdle-models" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Return coefficients, standard errors, etc. Calculate and return residuals from models.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="se.coef.html">se.coef()</a></code> </p>
        </td>
        <td><p>Return coefficient standard errors</p></td>
      </tr><tr><td>
          <p><code><a href="ZlmFit-class.html">lrTest(<i>&lt;ZlmFit&gt;</i>,<i>&lt;CoefficientHypothesis&gt;</i>)</a></code> <code><a href="ZlmFit-class.html">lrTest(<i>&lt;ZlmFit&gt;</i>,<i>&lt;Hypothesis&gt;</i>)</a></code> <code><a href="ZlmFit-class.html">lrTest(<i>&lt;ZlmFit&gt;</i>,<i>&lt;matrix&gt;</i>)</a></code> <code><a href="ZlmFit-class.html">waldTest(<i>&lt;ZlmFit&gt;</i>,<i>&lt;CoefficientHypothesis&gt;</i>)</a></code> <code><a href="ZlmFit-class.html">waldTest(<i>&lt;ZlmFit&gt;</i>,<i>&lt;Hypothesis&gt;</i>)</a></code> <code><a href="ZlmFit-class.html">coef(<i>&lt;ZlmFit&gt;</i>)</a></code> <code><a href="ZlmFit-class.html">vcov(<i>&lt;ZlmFit&gt;</i>)</a></code> <code><a href="ZlmFit-class.html">se.coef(<i>&lt;ZlmFit&gt;</i>)</a></code> </p>
        </td>
        <td><p>An S4 class to hold the output of a call to zlm</p></td>
      </tr><tr><td>
          <p><code><a href="predict.ZlmFit.html">predict(<i>&lt;ZlmFit&gt;</i>)</a></code> </p>
        </td>
        <td><p>Return predictions from a ZlmFit object.</p></td>
      </tr><tr><td>
          <p><code><a href="logFC.html">logFC()</a></code> <code><a href="logFC.html">getLogFC()</a></code> </p>
        </td>
        <td><p>Calculate log-fold changes from hurdle model components</p></td>
      </tr><tr><td>
          <p><code><a href="collectResiduals.html">collectResiduals()</a></code> <code><a href="collectResiduals.html">discrete_residuals_hook()</a></code> <code><a href="collectResiduals.html">continuous_residuals_hook()</a></code> <code><a href="collectResiduals.html">combined_residuals_hook()</a></code> <code><a href="collectResiduals.html">deviance_residuals_hook()</a></code> <code><a href="collectResiduals.html">fitted_phat()</a></code> <code><a href="collectResiduals.html">partialScore()</a></code> </p>
        </td>
        <td><p>Residual hooks and collection methods</p></td>
      </tr><tr><td>
          <p><code><a href="impute.html">impute()</a></code> </p>
        </td>
        <td><p>impute missing continuous expression for plotting</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="gene-set-enrichment-testing">Gene set enrichment testing <a href="#gene-set-enrichment-testing" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Tests on the average differential expression effect in a gene set, accounting for gene-gene correlations. Uses bootstraps to assess the gene-gene correlations.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="gseaAfterBoot.html">gseaAfterBoot()</a></code> <code><a href="gseaAfterBoot.html">gsea_control()</a></code> </p>
        </td>
        <td><p>Gene set analysis for hurdle model</p></td>
      </tr><tr><td>
          <p><code><a href="bootVcov1.html">pbootVcov1()</a></code> <code><a href="bootVcov1.html">bootVcov1()</a></code> </p>
        </td>
        <td><p>Bootstrap a zlmfit</p></td>
      </tr><tr><td>
          <p><code><a href="summary-GSEATests-method.html">summary(<i>&lt;GSEATests&gt;</i>)</a></code> </p>
        </td>
        <td><p>Summarize gene set enrichment tests</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="plotting-methods">Plotting methods <a href="#plotting-methods" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Miscellaneous plotting methods. Concordance, PCA biplots, and effects plots for fitted hurdle models.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="myBiplot.html">myBiplot()</a></code> </p>
        </td>
        <td><p>Makes a nice BiPlot</p></td>
      </tr><tr><td>
          <p><code><a href="plotSCAConcordance.html">plotSCAConcordance()</a></code> </p>
        </td>
        <td><p>Concordance plots of filtered single vs n-cell assays</p></td>
      </tr><tr><td>
          <p><code><a href="stat_ell.html">stat_ell()</a></code> </p>
        </td>
        <td><p>Plot confidence ellipse in 2D</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="legacy-methods">Legacy methods <a href="#legacy-methods" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p><p>Older and perhaps obsoleted methods.</p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="LRT.html">LRT()</a></code> </p>
        </td>
        <td><p>Likelihood Ratio Tests for SingleCellAssays</p></td>
      </tr><tr><td>
          <p><code><a href="computeEtFromCt.html">computeEtFromCt()</a></code> </p>
        </td>
        <td><p>Compute the Et from the Ct</p></td>
      </tr><tr><td>
          <p><code><a href="mast_filter.html">mast_filter()</a></code> <code><a href="mast_filter.html">burdenOfFiltering()</a></code> </p>
        </td>
        <td><p>Filter a SingleCellAssay</p></td>
      </tr><tr><td>
          <p><code><a href="expavg.html">expavg()</a></code> </p>
        </td>
        <td><p>Exponential average</p></td>
      </tr><tr><td>
          <p><code><a href="filterLowExpressedGenes.html">filterLowExpressedGenes()</a></code> </p>
        </td>
        <td><p>Filter low-expressing genes</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Andrew McDavid, Greg Finak, Masanao Yajima.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer></div>

  


  

  </body></html>

